File: src/at/vnr.ts
/**
 * VNR (Versicherungsnummer).
 *
 * Austrian social insurance number. 10 digits: 3-digit
 * serial + 1 check digit + 6-digit birth date (DDMMYY).
 * The check digit is computed using a weighted sum
 * mod 11 over the 9 non-check digits.
 *
 * The 6-digit "birth date" field is not always a real
 * calendar date. Sozialversicherung.at documents that
 * persons with unknown date of birth are issued
 * substitute values (01.01. or 01.07.), and that
 * months 13, 14, etc. are issued when the daily serial
 * pool for a given substitute date is exhausted. We
 * therefore do not enforce calendar validity on this
 * field — only the checksum is gating.
 *
 * @see https://de.wikipedia.org/wiki/Sozialversicherungsnummer#%C3%96sterreich
 * @see https://www.sozialversicherung.at/cdscontent/?contentid=10007.820902&viewmode=content
 */

import type { ValidateResult, Validator } from "../types";

import { clean } from "#util/clean";
import { randomDigits, randomInt } from "#util/generate";
import { err } from "#util/result";
import { isdigits } from "#util/strings";

/**
 * Weights for the 9 non-check digits (positions
 * 0,1,2, then 4,5,6,7,8,9). Check digit is at
 * position 3 and equals the weighted sum mod 11.
 */
const WEIGHTS = [3, 7, 9, 5, 8, 4, 2, 1, 6] as const;

const compact = (value: string): string =>
  clean(value, " -/").trim();

const validate = (value: string): ValidateResult => {
  const v = compact(value);

  if (v.length !== 10) {
    return err(
      "INVALID_LENGTH",
      "Austrian VNR must be 10 digits",
    );
  }

  if (!isdigits(v)) {
    return err(
      "INVALID_FORMAT",
      "Austrian VNR must contain only digits",
    );
  }

  // Per sozialversicherung.at, the 3-digit serial
  // (positions 0-2) never starts with zero.
  if (v[0] === "0") {
    return err(
      "INVALID_COMPONENT",
      "Austrian VNR serial must not start with zero",
    );
  }

  // Check digit at position 3: weighted sum of the
  // other 9 digits mod 11. If remainder is 10 the
  // number is invalid.
  const digits = v.slice(0, 3) + v.slice(4);
  let sum = 0;
  for (const [i, weight] of WEIGHTS.entries()) {
    sum += Number(digits[i]) * weight;
  }
  const check = sum % 11;
  if (check === 10) {
    return err(
      "INVALID_CHECKSUM",
      "Austrian VNR check digit is invalid (mod 11 " +
        "remainder is 10)",
    );
  }
  if (check !== Number(v[3])) {
    return err(
      "INVALID_CHECKSUM",
      "Austrian VNR check digit mismatch",
    );
  }

  return { valid: true, compact: v };
};

const format = (value: string): string => {
  const v = compact(value);
  return `${v.slice(0, 4)} ${v.slice(4, 6)}${v.slice(6, 8)}${v.slice(8)}`;
};

/** Generate a random valid Austrian VNR. */
const generate = (): string => {
  for (;;) {
    const serial = randomDigits(3);
    const day = String(randomInt(1, 28)).padStart(2, "0");
    const month = String(randomInt(1, 12)).padStart(2, "0");
    const year = String(randomInt(0, 99)).padStart(2, "0");
    const digits = `${serial}${day}${month}${year}`;

    let sum = 0;
    for (const [i, weight] of WEIGHTS.entries()) {
      sum += Number(digits[i]) * weight;
    }

    const check = sum % 11;
    if (check === 10) continue;
    return `${serial}${String(check)}${day}${month}${year}`;
  }
};

/** Austrian Social Insurance Number. */
const vnr: Validator = {
  name: "Austrian Social Insurance Number",
  localName: "Versicherungsnummer",
  abbreviation: "VNR",
  aliases: [
    "VNR",
    "SVNR",
    "Versicherungsnummer",
    "Sozialversicherungsnummer",
  ] as const,
  candidatePattern: "\\d{4}\\s?\\d{6}",
  country: "AT",
  entityType: "person",
  sourceUrl:
    "https://de.wikipedia.org/wiki/Sozialversicherungsnummer",
  lengths: [10] as const,
  examples: ["1237010180"] as const,
  compact,
  format,
  validate,
  generate,
};

export default vnr;
export { compact, format, generate, validate };

File: crates/stdnum-core/src/validators/at/vnr.rs
//! Austrian Social Insurance Number.

use crate::{
  ValidationResult, Validator,
  types::{CountryCode, EntityType, ValidatorScope, ValidatorSpec},
  validators::common::{
    compact as compact_identifier, digit, generate_from_examples, groups,
    invalid_checksum, invalid_component, require_digits, weighted_sum,
  },
};

const EXAMPLES: &[&str] = &["1237010180"];

pub static VALIDATOR: Validator = Validator::new(ValidatorSpec {
  id: "at.vnr",
  name: "Austrian Social Insurance Number",
  local_name: "Versicherungsnummer",
  abbreviation: "VNR",
  aliases: &[
    "VNR",
    "SVNR",
    "Versicherungsnummer",
    "Sozialversicherungsnummer",
  ],
  candidate_pattern: r"\d{4}\s?\d{6}",
  scope: ValidatorScope::Country(CountryCode::At),
  entity_type: EntityType::Person,
  source_url: Some("https://de.wikipedia.org/wiki/Sozialversicherungsnummer"),
  lengths: &[10],
  examples: EXAMPLES,
  compact,
  format,
  validate,
  generate: Some(generate),
  parse: None,
});

#[must_use]
pub fn compact(value: &str) -> String {
  compact_identifier(value)
}

#[must_use]
pub fn format(value: &str) -> String {
  groups(&compact(value), &[(0, 4), (4, 10)], " ")
}

pub fn validate(value: &str) -> ValidationResult {
  let value = compact(value);
  require_digits(&value, 10)?;
  let day = value.get(4..6).and_then(|value| value.parse::<u32>().ok());
  if value.starts_with('0') || !day.is_some_and(|day| (1..=31).contains(&day)) {
    return Err(invalid_component());
  }
  let payload = format!(
    "{}{}",
    value.get(..3).unwrap_or(""),
    value.get(4..).unwrap_or("")
  );
  let check = weighted_sum(&payload, &[3, 7, 9, 5, 8, 4, 2, 1, 6]) % 11;
  if check == 10 || check != digit(&value, 3) {
    return Err(invalid_checksum());
  }
  Ok(value)
}

/// Validate the exact output of [`compact`] without allocating.
#[must_use]
pub fn is_valid_canonical(value: &str) -> bool {
  let Ok(bytes) = <&[u8; 10]>::try_from(value.as_bytes()) else {
    return false;
  };
  let &[d0, d1, d2, check_digit, day0, day1, d6, d7, d8, d9] = bytes;
  if !bytes.iter().all(u8::is_ascii_digit) || d0 == b'0' {
    return false;
  }
  let day = u32::from(day0.saturating_sub(b'0'))
    .saturating_mul(10)
    .saturating_add(u32::from(day1.saturating_sub(b'0')));
  if !(1..=31).contains(&day) {
    return false;
  }
  let payload = [d0, d1, d2, day0, day1, d6, d7, d8, d9];
  let check = payload.iter().zip([3_u32, 7, 9, 5, 8, 4, 2, 1, 6]).fold(
    0_u32,
    |sum, (byte, weight)| {
      sum.saturating_add(
        u32::from(byte.saturating_sub(b'0')).saturating_mul(weight),
      )
    },
  ) % 11;
  check != 10 && u32::from(check_digit.saturating_sub(b'0')) == check
}

#[must_use]
pub fn generate() -> String {
  generate_from_examples(EXAMPLES, compact, validate)
}
